### PR TITLE
add initial skeleton and logic for buildkite k8s/gke-backed terraform module

### PR DESCRIPTION
**Pull request by 0x0I**
This PR introduces infrastructure components for provisioning **K8s-backed Buildkite agent clusters**, supporting both *containerized* and *non-containerized* Job execution profiles and uploads/downloads of job artifacts to *O(1) Lab's Google Cloud Storage*.

**Notes:**
- required variable inputs consist of cluster ID and version control settings and can/generally should be expressed as environment variables within an operator's execution environment
  * e.g. rather than specify `agent_token` as a Terraform input variable, set `$TF_VAR_agent_token=` to the appropriate value instead (prior to running `terraform ...`)

See *prototype* running as single-node (aka my work laptop) mini-kube cluster: [here](https://buildkite.com/organizations/o-1-labs-2/agents)

**References:**
- Buildkite agent configuration [settings](https://buildkite.com/docs/agent/v3/configuration#configuration-settings)
- Agent VCS SSH [options](https://buildkite.com/docs/agent/v3/github-ssh-keys) 
- Buildkite Helm Chart [README](https://github.com/buildkite/charts/blob/master/stable/agent/README.md)

**Fixes:**
- Coda issue [#4803](https://github.com/CodaProtocol/coda/issues/4803)
- Coda issue [#4804](https://github.com/CodaProtocol/coda/issues/4804)

### CODE REVIEW

**Comment by 0x0I** [comment]
@yourbuddyconner @bkase Not sure if it's worth creating an RFC for discussing how we can about setting up test-infra and testing infra components like this and others...

**Comment by 0x0I** [comment]
Optionally and safely CONDITIONALLY create Terraform resources: https://ilhicas.com/2019/08/20/Second-Conditional-terraform-resouce-for-each.html

Could come in handy for avoiding `terraform apply`ing resources scoped for PROD/Live environments when operating in dev and CI/CD.

File: terraform/buildkite/example/main.tf [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        encrypt = true
+        region  = "us-west-2"
+        bucket  = "o1labs-terraform-state"
+        acl     = "bucket-owner-full-control"
+    }
+}
+
+# Required input variables -- recommended to express as environment vars (e.g. TF_VAR_***)
+variable "agent_token" {}
+variable "agent_vcs_privkey" {}
```
Comment (bkase) [inline terraform/buildkite/example/main.tf:14]: where does this come from?

Comment (yourbuddyconner) [reply terraform/buildkite/example/main.tf:14]: This is likely referring to a deploy key for Github

Comment (0x0I) [reply terraform/buildkite/example/main.tf:14]: hey @bkase @yourbuddyconner, right - the `agent_vcs_privkey` refers to a GIthub SSH access private key (or machine deploy key for CI/CD, bots, etc).

Also, both it and the `agent_token` should be set as ENV VARS prior to applying. For example:

```bash
export TF_VAR_AGENT_TOKEN="<buildkite-agent-access-token>"
export TF_VAR_AGENT_VCS_PRIVKEY="$(cat ~/.ssh/id_rsa)
```

...thinking it should make it pretty simple for infra/dev use and CI/CD pipelines + even bots.

Comment (0x0I) [reply terraform/buildkite/example/main.tf:14]: @bkase - yep, mostly all done. Just working on experimenting with different test frameworks, etc validation related.

**Comment by 0x0I** [comment]
> Ah, one point of configurability that would be nice here is to specify the number of "small" and "large" nodes. Which would then go and provision those instances and apply the right tags. We should probably also tag all of these "cluster=" or something as well.
> 
> This can come in a follow up PR of course.

Got it and for sure, sounds good! So the thinking here is that the level of configurability you're referring to pretty much already exists at the cluster definition level. For example, we could create a cluster now to match the "small" and "large" node cluster topology with something like the following:

```bash
$ cat .../terraform/buildkite/small-large-sizing-cluster.tf

...

# Main resource entrypoint
module "buildkite_small" {
    source                  = "../../modules/kubernetes/buildkite-agent"

    google_app_credentials  =   var.google_credentials
    k8s_cluster_name           =   "coda-infra-east"
    k8s_cluster_region          =   "us-east1"
    k8s_provider                    =   var.k8s_provider

    cluster_name                   =   "buildkite-small"
    cluster_namespace         =   "buildkite-small"
    agent_token                     =   var.agent_token
    agent_vcs_privkey           =   var.agent_vcs_privkey
    agent_meta                      =   "queue=buildkite-small,size=small"
    num_agents                     =   5
    agent_resources =  {
        limits = {
           cpu                            = 100m
           memory                     = 1G
        }
     }
     image = var.coda_image
     seedPeers = concat(var.additional_seed_peers, local.seed_peers)
    }
  #  memory: 1024Mi
  # requests:
  #  cpu: 100m
  #  memory: 128Mi
}

module "buildkite_large" {
    source                  = "../../modules/kubernetes/buildkite-agent"

    google_app_credentials  =   var.google_credentials
    k8s_cluster_name        =   "coda-infra-east"
    k8s_cluster_region      =   "us-east1"
    k8s_provider            =   var.k8s_provider

    cluster_name            =   "example"
    cluster_namespace       =   "example"
    agent_token             =   var.agent_token
    agent_vcs_privkey       =   var.agent_vcs_privkey
    agent_meta              =   "queue=buildkite-large,size=large"
    num_agents              =   10
    agent_resources =  {
        limits = {
           cpu                            = 1g
           memory                     = 10G
        }
     }
}

**Comment by 0x0I** [comment]
^^^updated the cluster spec based on *size vs. large* sizing using TF's [foreach](https://www.terraform.io/docs/configuration/resources.html#the-each-object) - could be pretty solid for parameterizing(?) cluster provisioning.

**Comment by yourbuddyconner** [comment]
This looks great, @0x0I could you please run `terraform fmt` on the buildkite folder, then probably g2g!

**Comment by 0x0I** [comment]
> This looks great, @0x0I could you please run `terraform fmt` on the buildkite folder, then probably g2g!

Cool, got it and done @yourbuddyconner!

**Pull request merged**